cpython/wasm_assets.py

```diff
@@ -0,0 +1,233 @@
+    "venv/",
+    # build system
+    "lib2to3/",
+    # other platforms
+    "_osx_support.py",
+    # Not supported by browser
+    "curses/",
+    "dbm/",
+    "idlelib/",
+    "tkinter/",
+    "turtle.py",
+    "turtledemo",
+    "webbrowser.py",
+)
+
+# This files are unvendored from the stdlib by default
+UNVENDORED_FILES = (
+    "test/",
+    "distutils/",
+    "sqlite3",
+    "ssl.py",
+    "lzma.py",
+)
+
+# TODO: These modules have test directory which we unvendors it separately.
+#       So we should not pack them into the zip file in order to make e.g. import ctypes.test work.
+#       Note that all these tests are moved to the subdirectory of `test` module in upstream CPython 3.12.0a1.
+#       So we don't need this after we upgrade to 3.12.0
+NOT_ZIPPED_FILES = ("ctypes/", "unittest/")
+
+
+class PyZipFileShortTraceback(zipfile.PyZipFile):
+    """
+    This class is for creating .pyc files with shortened traceback.
+
+    With PyZipFile:
+
+    Traceback (most recent call last):
+        File "/src/cpython/installs/python-3.10.2/lib/python3.10/pathlib.py", line 1152, in write_text
+        File "/src/cpython/installs/python-3.10.2/lib/python3.10/pathlib.py", line 1117, in open
+
+    With this class:
+
+    Traceback (most recent call last):
+        File "pathlib.py", line 1152, in write_text
+        File "pathlib.py", line 1117, in open
+    """
+
+    def __init__(
+        self,
+        file,
+        mode="r",
+        compression=zipfile.ZIP_STORED,
+        allowZip64=True,
+        optimize=-1,
+        basedir=None,
+    ):
+        super().__init__(file, mode, compression, allowZip64, optimize)
+        self.pyc_dir = Path(tempfile.mkdtemp())
+        self.base_dir = basedir
+
+    def __del__(self):
+        super().__del__()
+        shutil.rmtree(self.pyc_dir)
+
+    def _get_codename(self, pathname, basename):
+        """Return (filename, archivename) for the path.
+        Given a module name path, return the correct file path and
+        archive name, compiling if necessary.  For example, given
+        /python/lib/string, return (/python/lib/string.pyc, string).
+        """
+
+        def _compile(file, optimize=-1):
+            import py_compile
+
+            if self.debug:
+                print("Compiling", file)
+
+            rel_path = Path(file).relative_to(self.base_dir)
+            compile_path_full = self.pyc_dir / rel_path.with_suffix(".pyc")
+            py_compile.compile(
+                file,
+                cfile=str(compile_path_full),
+                dfile=str(rel_path),
+                doraise=True,
+                optimize=optimize,
+            )
+
+            return compile_path_full
+
+        file_py = pathname + ".py"
+        file_pyc = _compile(file_py, optimize=self._optimize)
+
+        archivename = os.path.split(file_pyc)[1]
+        if basename:
+            archivename = f"{basename}/{archivename}"
+
+        return (file_pyc, archivename)
+
+
+def create_stdlib_zip(
```

Comment:
Maybe we could put most of these functions in pyodide_build.packer (or something similar) so we could unit test them better? E.g. running this on some .py files in a temp dir, and checking the output zip (also to check that it doesn't contain .py).

Reply:
Sound good to me. That would require some modification in our build system since pyodide-build is not used for CPython build steps now, but I think it wouldn't be that hard.